src/backend/distributed/Makefile

```diff
@@ -39,9 +37,12 @@ all:
 NO_PGXS = 1
 
 SHLIB_LINK = $(libpq)
-
+SHLIB_LINK_INTERNAL += -L../columnar -l:citus_columnar$(DLSUFFIX)
```

Comment:
Assuming this is the extra change added, I have tried compiling on OSX with this configuration and it didn't work.

When looking at the [linker man page](http://www.manpagez.com/man/1/ld64/) we find the following:
```
Options that control libraries
     -lx         This option tells the linker to search for libx.dylib or
                 libx.a in the library search path.  If string x is of the
                 form y.o, then that file is searched for in the same places,
                 but without prepending `lib' or appending `.a' or `.dylib' to
                 the filename.
```

With `$(DLSUFFIX)` added osx provides the flag as follows: `-l:citus_columnar.so` which corresponds to the `.so` file compiled from the columnar submodule. However as stated in the man page quoted, it will _only_ resolve the file as is, when it ends in `.o`. By linking against `citus_columnar.so` it tries to resolve `libcitus_columnar.so.dylib`.

I have tried experimenting a bit with the variations what we could pass to the linker and which files are present. For example, we can make the linker succeed if we have created the following symbolic link in the `src/backend/columnar` directory: `ln -s citus_columnar.so libcitus_columnar.so.dylib`. However, this only solves the linking phase. Now when we create a cluster it crashes with the following error:

```
2022-03-03 12:37:42.793 CET [87561] FATAL:  could not load library "/Users/nilsdijk/.pgenv/pgsql-14.1/lib/citus.so": dlopen(/Users/nilsdijk/.pgenv/pgsql-14.1/lib/citus.so, 0x000A): symbol not found in flat namespace '_ColumnarSupportsIndexAM'
2022-03-03 12:37:42.794 CET [87561] LOG:  database system is shut down
```

I get the same error if I remove the `-L`/`-l` flags from the linker, and instead hardcode the expected missing symbols like the following:
```
SHLIB_LINK_INTERNAL += -Wl,-U,_ColumnarSupportsIndexAM
SHLIB_LINK_INTERNAL += -Wl,-U,_ColumnarTableSetOptions_hook
SHLIB_LINK_INTERNAL += -Wl,-U,_CompressionTypeStr
SHLIB_LINK_INTERNAL += -Wl,-U,_IsColumnarTableAmTable
SHLIB_LINK_INTERNAL += -Wl,-U,_ReadColumnarOptions
```

With having two ways of making compilation pass, I tried focussing on how we could make the. loading of the citus library work in postgres.

Adding `citus_columnar` looked like an option to me, however, adding it after `citus` keeps the same error. Putting it before citus makes citus error out due to it _not_ being the first preload library.

How does the loading of the library work on linux when `citus.so` is linked to `citus_columnar.so` and loaded without its symbols being available yet? OSX shows the error above, even when we link with the `-weak-l` flag, which should be ok with missing symbols at runtime:

``` 
-weak-lx    This is the same as the -lx but forces the library and all
                 references to it to be marked as weak imports.  That is, the
                 library is allowed to be missing at runtime.
```

I think if we can understand how linux's dlopen works compared to osx's we might be able to find a way to link on both platforms.

Also, have we looked at how postgis does this? I believe they come with many different `.so`'s. Albeit getting it to work on osx is quite a challenge, due to arcane versions of dependency libraries, it does work on OSX. I can't imagine they don't have links between their submodules.

Reply:
Ugh. Mac does not make this easy. See [here](https://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-macos/2339910#2339910):

> It is not possible to link against bundles as if they were shared libraries. However, it is possible that a bundle is linked against real shared libraries; those will be loaded automatically when the bundle is loaded.

Our approach on linux was to do exactly that: have a loadable module that could also be linked against. To achieve the same goal on Mac perhaps we'd need a shared library for columnar (`citus_columnar.dylib`) that's shared between two different bundles (`citus.so` and `citus_columnar.so`).

The reason we want the library to actually be shared (rather than statically linked in two independent `.so` files) is that there's a hook in columnar for setting options that citus needs to set (so that it can propagate table option changes). The reason we need to do that is because normal table options ("reloptions") aren't extensible, so we use a UDF. And it's hard to hook the calling of a UDF. My, how the smallest engineering compromises snowball into a mess.

Re-thinking this from the original problem: maybe we should capture `ALTER TABLE` with a `ProcessUtility_hook` in columnar, and then extract the columnar options from that. Then citus can have its own hook to do essentially the same thing, but at a cluster level (it already has that, but might require a tweak). The rest of the shared symbols are "just" code, and therefore statically linking should be functionally the same as dynamically linking, and we can simplify the build. This would impact the user experience for columnar, but in a good way.

Reply:
if table options are the only reason we would want to link citus to the columnar binaries I think this makes a lot of sense.

> This would impact the user experience for columnar, but in a good way
I completely agree here.

The final destination should be that postgres has native support for custom options on a table am. Citus would then intercept the `ALTER TABLE` commands here and use the public interface postgres provides for the options to understand what to do for table access methods.

There are two parts to this as I can currently see.

 - changing table options on already distributed columnar tables
 - distributing a columnar table that has non-default options set.

For the first there is a balancing act between the order of utility hooks called. If Citus is called first we could simply read the columnar options for distributed tables. However, since Citus is normally the first extension loaded citus' utility hook will be called as last. Meaning, columnar should have already stripped the columnar specific options from the `ALTER TABLE ... SET ...` statement. I don't exactly understand how we could solve this in a way that feels idiomatic.

For the second case I don't even know where to start without linking. We almost need to understand and replicate the logic for reading the options from the catalog.

@jeff-davis how do you envision the implementation of `ALTER TABLE .... SET ....` to integrate both seemless within Columnar _and_ Citus?

Reply:
1. **changing table options on already distributed columnar table** - if the dependency arrow goes citus -> columnar, then doesn't it make sense to load columnar first? I don't think it interferes with anything citus is doing, and citus could check that `ProcessUtility_hook == ColumnarProcessUtilityHook` for error checking. In other words, can we just move the `load_file()` to the top of the citus `_PG_init()`?
2. **distributing a columnar table that has non-default options set** - I believe this problem is unchanged, and is already solved in `alter_table.c:ConvertTable()` and `GetShardedTableDDLCommandColumnar()`. Did I miss something?

Reply:
> and citus could check that `ProcessUtility_hook == ColumnarProcessUtilityHook` for error checking
wouldn't this still require Citus to link to Columnar?

Reply:
Correct again, my mistake. Worst case, we just parse `shared_preload_libraries` to make sure it doesn't have any other extension before citus and citus_columnar, and that could be a reasonable error check.

Then if we fix the linking issue then we can improve the checks. I think it's still good to fix the linking issue, by the way, I just like the idea of making `ALTER TABLE` work too.

Reply:
Definitely agree on the `ALTER TABLE ... SET` being a positive change! Love it!

Parsing `shared_preload_libraries` approach sounds reasonable to me.
I think Citus needs some work in its utility hook to call into a stored pointer instead of the standard util hook function of postgres. I don't think that has been taken into account just yet. But I agree that it will be definitely easier when citus calls into the columnar hook.